lib/auth/iam_join.go

```diff
@@ -0,0 +1,301 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package auth
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"crypto/rand"
+	"encoding/base64"
+	"encoding/json"
+	"io"
+	"net/http"
+	"net/url"
+	"regexp"
+	"strings"
+
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/api/utils"
+
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/sts"
+	"github.com/gravitational/trace"
+	"google.golang.org/grpc/peer"
+)
+
+const (
+	expectedSTSIdentityRequestBody = "Action=GetCallerIdentity&Version=2011-06-15"
+	stsHost                        = "sts.amazonaws.com"
+	challengeHeaderKey             = "X-Teleport-Challenge"
+	normalizedChallengeHeaderKey   = "x-teleport-challenge"
+	authHeaderKey                  = "Authorization"
+)
+
+var signedHeadersRe = regexp.MustCompile(`^AWS4-HMAC-SHA256 Credential=\S+, SignedHeaders=(\S+), Signature=\S+$`)
+
+func validateSTSIdentityRequest(req *http.Request, challenge string) error {
+	if req.Host != stsHost {
+		return trace.AccessDenied("sts identity request is for unknown host %q", req.Host)
+	}
+
+	if req.Method != http.MethodPost {
+		return trace.AccessDenied("sts identity request method %q does not match expected method %q", req.RequestURI, http.MethodPost)
+	}
+
+	if req.Header.Get(challengeHeaderKey) != challenge {
+		return trace.AccessDenied("sts identity request does not include challenge header or it does not match")
+	}
+
+	authHeader := req.Header.Get(authHeaderKey)
+	matches := signedHeadersRe.FindStringSubmatch(authHeader)
+	// first match should be the full header, second is the SignedHeaders
+	if len(matches) < 2 {
+		return trace.AccessDenied("sts identity request Authorization header is invalid")
+	}
+	signedHeadersString := matches[1]
+	signedHeaders := strings.Split(signedHeadersString, ";")
```

Comment:
It might be worth a quick note about what `signedHeaders` is expected to contain, as I was quite confused until I cross-referenced with [this](https://docs.aws.amazon.com/general/latest/gr/sigv4-signed-request-examples.html) - maybe part of the function-level discussion I suggested above?

Reply:
added the example including `signedHeaders` above